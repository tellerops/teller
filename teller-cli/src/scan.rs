use comfy_table::presets::NOTHING;
use comfy_table::{Cell, Table};
use eyre::Result;
use teller_core::{scan, teller::Teller};

use crate::cli::ScanArgs;
use crate::Response;

fn hide_chars(s: &str) -> String {
    let mut result = String::new();
    let chars_to_display = s.chars().take(2).collect::<String>();
    let asterisks = "*".repeat(3);
    result.push_str(&chars_to_display);
    result.push_str(&asterisks);
    result
}

/// Scan a folder for secrets fetched from providers
///
/// # Errors
///
/// This function will return an error if the operation fails
#[allow(clippy::future_not_send)]
pub async fn run(teller: &Teller, args: &ScanArgs) -> Result<Response> {
    let opts = scan::Opts {
        include_all: args.all,
        include_binary: args.binary,
    };

    let kvs = teller.collect().await?;
    let res = teller.scan(&args.root, &kvs, &opts)?;
    let count = res.len();
    eprintln!("scanning for {} item(s) in {}", kvs.len(), args.root);
    if args.json {
        println!("{}", serde_json::to_string_pretty(&res)?);
    } else {
        let mut table = Table::new();
        table.load_preset(NOTHING);
        for m in res {
            let pos = m.position.unwrap_or((0, 0));
            table.add_row(vec![
                Cell::new(format!("{}:{}", pos.0, pos.1)),
                Cell::new(m.path.to_string_lossy()),
                Cell::new(hide_chars(&m.query.value)),
                Cell::new(
                    m.query
                        .provider
                        .map_or_else(|| "n/a".to_string(), |p| p.kind.to_string())
                        .to_string(),
                ),
                Cell::new(m.query.path.map_or_else(|| "n/a".to_string(), |p| p.path)),
            ]);
        }
        println!("{table}");
    }
    eprintln!("found {count} result(s)");

    if args.error_if_found && count > 0 {
        Response::fail()
    } else {
        Response::ok()
    }
}
