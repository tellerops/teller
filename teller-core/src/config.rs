use std::cmp::Ordering;
use std::path::PathBuf;
use std::{
    collections::{BTreeMap, HashMap},
    path::Path,
};

use fs_err as fs;
use serde_derive::{Deserialize, Serialize};
use teller_providers::config::{PathMap, ProviderCfg, KV};
use teller_providers::providers::ProviderKind;
use tera::{Context, Tera};

use crate::Result;

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
pub struct Config {
    pub providers: BTreeMap<String, ProviderCfg>,
}

#[derive(Serialize)]
pub struct RenderTemplate {
    pub providers: Vec<ProviderKind>,
}

fn apply_eqeq(config: &mut Config) {
    config.providers.iter_mut().for_each(|(_name, provider)| {
        provider.maps.iter_mut().for_each(|pm| {
            pm.keys.iter_mut().for_each(|(k, v)| {
                // THINK: replace with:
                // 1. templating: {{id}} (identity), {{snake_case}} (snake case it)
                // 2. other symbols: == id, ^^ capitalize, snake case __ lower snake case
                if v == "==" {
                    v.clone_from(k);
                }
            });
        });
    });
}

impl Config {
    /// Config from text
    ///
    /// # Errors
    ///
    /// This function will return an error if serialization fails
    pub fn with_vars(text: &str, vars: &HashMap<String, String>) -> Result<Self> {
        let rendered_text = Tera::one_off(text, &Context::from_serialize(vars)?, false)?;
        let mut config: Self = serde_yaml::from_str(&rendered_text)?;

        apply_eqeq(&mut config);

        Ok(config)
    }

    /// Config from text
    ///
    /// # Errors
    ///
    /// This function will return an error if serialization fails
    pub fn from_text(text: &str) -> Result<Self> {
        Self::with_vars(text, &HashMap::new())
    }

    /// Config from file
    ///
    /// # Errors
    ///
    /// This function will return an error if IO fails
    pub fn from_path(path: &Path) -> Result<Self> {
        Self::from_text(&fs::read_to_string(path)?)
    }

    /// Create configuration template file
    ///
    /// # Errors
    /// When could not convert config to string
    pub fn render_template(data: &RenderTemplate) -> Result<String> {
        let res: BTreeMap<String, ProviderCfg> = data
            .providers
            .iter()
            .map(|p| {
                (
                    format!("{p}_1"),
                    ProviderCfg {
                        kind: p.clone(),
                        maps: vec![PathMap::from_path("example/dev")],
                        ..ProviderCfg::default()
                    },
                )
            })
            .collect();

        let config = Self { providers: res };

        let a: String = serde_yaml::to_string(&config)?;
        Ok(a)
    }
}

#[derive(Debug, Clone, Serialize, Eq, PartialEq)]
pub struct Match {
    pub path: PathBuf,
    pub position: Option<(usize, usize)>,
    pub offset: usize,
    pub query: KV,
}

impl PartialOrd for Match {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Match {
    fn cmp(&self, other: &Self) -> Ordering {
        let query_cmp = self.query.cmp(&other.query);

        if query_cmp != Ordering::Equal {
            return query_cmp;
        }

        self.offset.cmp(&other.offset)
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_yaml_snapshot;

    use super::*;
    #[test]
    fn load_config() {
        std::env::set_var("TEST_LOAD_1", "DEV");
        let config = Config::from_path(Path::new("fixtures/config.yml")).unwrap();
        assert_eq!(config.providers.len(), 2);
        assert_yaml_snapshot!(config);
    }

    #[test]
    fn can_render_template_config() {
        let data = RenderTemplate {
            providers: vec![ProviderKind::Inmem, ProviderKind::Dotenv],
        };

        let config = Config::render_template(&data).unwrap();
        assert_yaml_snapshot!(config);
    }
}
