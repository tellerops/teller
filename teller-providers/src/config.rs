use std::cmp::Ordering;
use std::collections::BTreeMap;

use serde_derive::{Deserialize, Serialize};

use crate::providers::ProviderKind;

fn is_default<T: Default + PartialEq>(t: &T) -> bool {
    t == &T::default()
}

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
pub struct ProviderCfg {
    #[serde(rename = "kind")]
    pub kind: ProviderKind,
    #[serde(rename = "options", skip_serializing_if = "Option::is_none")]
    pub options: Option<serde_json::Value>,
    #[serde(rename = "name", skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    pub maps: Vec<PathMap>,
}

#[derive(Serialize, Deserialize, Debug, Clone, Default, Eq, PartialEq)]
pub enum Sensitivity {
    #[default]
    None,
    Low,
    Medium,
    High,
    Critical,
}

#[derive(Serialize, Deserialize, Debug, Clone, Default, Eq, PartialEq)]
pub struct ProviderInfo {
    pub kind: ProviderKind,
    pub name: String,
}
#[derive(Serialize, Deserialize, Debug, Clone, Default, Eq, PartialEq)]
pub struct PathInfo {
    pub id: String,
    pub path: String,
}

#[derive(Serialize, Deserialize, Debug, Clone, Default, Eq, PartialEq)]
pub struct MetaInfo {
    pub sensitivity: Sensitivity,
    pub redact_with: Option<String>,
    pub source: Option<String>,
    pub sink: Option<String>,
}
#[derive(Serialize, Deserialize, Debug, Clone, Default, Eq, PartialEq)]
pub struct KV {
    pub value: String,
    pub key: String, // mapped-to key
    pub from_key: String,
    pub path: Option<PathInfo>, // always toplevel
    pub provider: Option<ProviderInfo>,
    pub meta: Option<MetaInfo>,
}

impl PartialOrd for KV {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for KV {
    fn cmp(&self, other: &Self) -> Ordering {
        if let (Some(provider), Some(other_provider)) = (&self.provider, &other.provider) {
            let provider_cmp = provider.kind.cmp(&other_provider.kind);
            if provider_cmp != Ordering::Equal {
                return provider_cmp;
            }
        }

        self.key.cmp(&other.key)
    }
}

impl KV {
    #[must_use]
    pub fn to_data(kvs: &[Self]) -> BTreeMap<String, String> {
        let mut data = BTreeMap::new();
        for kv in kvs {
            data.insert(kv.key.clone(), kv.value.clone());
        }
        data
    }

    #[must_use]
    pub fn from_data(
        data: &BTreeMap<String, String>,
        pm: &PathMap,
        provider: &ProviderInfo,
    ) -> Vec<Self> {
        // map all of the data found
        if pm.keys.is_empty() {
            data.iter()
                .map(|(k, v)| Self::from_value(v, k, k, pm, provider.clone()))
                .collect::<Vec<_>>()
        } else {
            // selectively map only keys from pathmap
            pm.keys
                .iter()
                .filter_map(|(from_key, to_key)| {
                    data.get(from_key).map(|found_val| {
                        Self::from_value(found_val, from_key, to_key, pm, provider.clone())
                    })
                })
                .collect::<Vec<_>>()
        }
    }
    #[must_use]
    pub fn from_value(
        found_val: &str,
        from_key: &str,
        to_key: &str,
        pm: &PathMap,
        provider: ProviderInfo,
    ) -> Self {
        Self {
            value: found_val.to_string(),
            key: to_key.to_string(),
            from_key: from_key.to_string(),
            path: Some(PathInfo {
                path: pm.path.clone(),
                id: pm.id.to_string(),
            }),
            provider: Some(provider),
            meta: Some(MetaInfo {
                sensitivity: pm.sensitivity.clone(),
                redact_with: pm.redact_with.clone(),
                source: pm.source.clone(),
                sink: pm.sink.clone(),
            }),
        }
    }
    #[must_use]
    pub fn from_literal(path: &str, key: &str, value: &str, provider: ProviderInfo) -> Self {
        Self {
            value: value.to_string(),
            key: key.to_string(),
            from_key: key.to_string(),
            path: Some(PathInfo {
                id: path.to_string(),
                path: path.to_string(),
            }),
            provider: Some(provider),
            ..Default::default()
        }
    }

    /// represents a KV without any source (e.g. created manually by a user, pending insert to
    /// one of the providers)
    #[must_use]
    pub fn from_kv(key: &str, value: &str) -> Self {
        Self {
            value: value.to_string(),
            key: key.to_string(),
            from_key: key.to_string(),
            ..Default::default()
        }
    }
}

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
pub struct PathMap {
    pub id: String,
    #[serde(rename = "protocol", skip_serializing_if = "Option::is_none")]
    pub protocol: Option<String>,
    #[serde(rename = "path")]
    pub path: String,
    #[serde(default, rename = "keys", skip_serializing_if = "is_default")]
    pub keys: BTreeMap<String, String>,
    #[serde(default, rename = "decrypt", skip_serializing_if = "is_default")]
    pub decrypt: bool,
    #[serde(default, rename = "sensitivity", skip_serializing_if = "is_default")]
    pub sensitivity: Sensitivity,
    #[serde(
        default,
        rename = "redact_with",
        skip_serializing_if = "Option::is_none"
    )]
    pub redact_with: Option<String>,
    #[serde(default, rename = "source", skip_serializing_if = "Option::is_none")]
    pub source: Option<String>,
    #[serde(default, rename = "sink", skip_serializing_if = "Option::is_none")]
    pub sink: Option<String>,
    // ignore population if optional + we got error
    #[serde(default, rename = "optional", skip_serializing_if = "is_default")]
    pub optional: bool,
}

impl PathMap {
    #[must_use]
    pub fn from_path(path: &str) -> Self {
        Self {
            path: path.to_string(),
            ..Default::default()
        }
    }
}
