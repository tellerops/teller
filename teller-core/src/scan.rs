use std::fs;

use aho_corasick::AhoCorasick;
use ignore::WalkBuilder;
use teller_providers::config::KV;
use unicode_width::UnicodeWidthStr;

use crate::{config::Match, io::is_binary_file, Error, Result};

#[derive(Debug, Clone, Default)]
pub struct Opts {
    pub include_all: bool,
    pub include_binary: bool,
}

/// (ln, col), 1 based (not zero based)
fn get_visual_position(text: &[u8], byte_position: usize) -> Option<(usize, usize)> {
    if byte_position >= text.len() || text.is_empty() {
        return None;
    }

    let lines = text
        .iter()
        .take(byte_position)
        .filter(|c| **c == b'\n')
        .count();
    let last_ln_start = text
        .iter()
        .take(byte_position)
        .rposition(|c| *c == b'\n')
        .unwrap_or(0);

    let len = UnicodeWidthStr::width(
        String::from_utf8_lossy(&text[last_ln_start..byte_position]).as_ref(),
    );

    // index starts from 1 for both
    Some((lines + 1, len + 1))
}

// aho
// offset into row/col visual https://github.com/zkat/miette/blob/f4d056e1ffeb9a0bf36e2a6501365bd7e00db22d/src/handlers/graphical.rs#L619
// match repr
///
/// # Errors
///
/// TODO
#[allow(clippy::module_name_repetitions)]
pub fn scan_root(root: &str, kvs: &[KV], opts: &Opts) -> Result<Vec<Match>> {
    let patterns = kvs.iter().map(|kv| kv.value.as_str()).collect::<Vec<_>>();
    let finder = AhoCorasick::new(patterns).map_err(|e| Error::Message(e.to_string()))?;

    let mut wb = WalkBuilder::new(root);

    let mut matches = vec![];
    for entry in wb
        .ignore(!opts.include_all)
        .git_ignore(!opts.include_all)
        .hidden(opts.include_all)
        .build()
        .filter_map(Result::ok)
        .filter(|ent| ent.path().is_file())
    {
        let path = entry.path();
        if is_binary_file(path)? && !opts.include_binary {
            continue;
        }

        let content = String::from_utf8_lossy(&fs::read(path)?).to_string();
        let bytes = content.as_bytes();

        finder.find_iter(&content).for_each(|aho_match| {
            matches.push(Match {
                path: path.to_path_buf(),
                query: kvs[aho_match.pattern()].clone(),
                position: get_visual_position(bytes, aho_match.start()),
                offset: aho_match.start(),
            });
        });
    }

    matches.sort();
    Ok(matches)
}

#[cfg(test)]
mod tests {
    use std::{
        fs,
        path::{Path, PathBuf},
    };

    use insta::assert_debug_snapshot;
    use teller_providers::{
        config::{ProviderInfo, KV},
        providers::ProviderKind,
    };

    use super::*;
    use crate::scan;

    fn normalize_path_separators(path: &Path) -> PathBuf {
        let path_str = path.to_string_lossy().replace('\\', "/");
        PathBuf::from(path_str)
    }
    fn normalize_matches(ms: &[Match]) -> Vec<Match> {
        ms.iter()
            .map(|m| Match {
                path: normalize_path_separators(&m.path),
                ..m.clone()
            })
            .collect::<Vec<_>>()
    }

    #[test]
    fn test_position() {
        assert_eq!(get_visual_position(b"", 4), None);
        assert_eq!(get_visual_position(b"", 1), None);
        assert_eq!(get_visual_position(b"", 0), None);
        assert_eq!(get_visual_position(b"a", 1), None);

        assert_eq!(get_visual_position(b"abcde\nfghi", 8), Some((2, 3)));
        assert_eq!(get_visual_position(b"abcde\r\nfghi", 8), Some((2, 2)));

        let text = r#"Ôï∏ 100% ‚ùØ j teller-rs
    /Users/jondot/spikes/teller-rs
    (base)
    ~/spikes/teller-rs on Ôêò master [!?] via ü¶Ä v1.73.0-nightly
    Ôï∏ 100% ‚ùØ code .
    (base)
    ~/spikes/teller-rs on Ôêò master [!?] via ü¶Ä v1.73.0-nightly
    Ôï∏ 100% ‚ùØ [WARN] - (starship::utils): Executing command "/opt/homebrew/bin/git" timed out.
    (base)
    ~/spikes/teller-rs on Ôêò master [!?] via ü¶Ä v1.73.0-nightly
    Ôï∏ 100% ‚ùØ open /Users/jondot/Movies
    (base)
    ~/spikes/teller-rs on Ôêò master [!?] via ü¶Ä v1.73.0-nightly
    Ôï∏ 100% ‚ùØ"#;
        let position = get_visual_position(text.as_bytes(), 438);
        assert_eq!(position, Some((11, 19)));
    }

    #[test]
    fn test_scan() {
        let provider = ProviderInfo {
            kind: ProviderKind::Inmem,
            name: "test".to_string(),
        };
        let kvs = vec![
            KV::from_literal("/some/path", "key1", "hashicorp", provider.clone()),
            KV::from_literal("/some/path", "key1", "dont-find-me", provider.clone()),
            KV::from_literal("/some/path", "key1", "trooper123", provider.clone()),
            KV::from_literal("/some/path", "key1", "pass1", provider.clone()),
            KV::from_literal("/some/path", "key1", "nested111", provider),
        ];

        let res = scan_root("fixtures", &kvs[..], &scan::Opts::default());
        assert_debug_snapshot!(normalize_matches(&res.unwrap()));

        let res = scan_root(
            "fixtures",
            &kvs[..],
            &scan::Opts {
                include_binary: true,
                include_all: false,
            },
        );
        assert_debug_snapshot!(normalize_matches(&res.unwrap()));

        fs::write("fixtures/git-ignored-file", "trooper123").expect("cannot write file");

        let res = scan_root(
            "fixtures",
            &kvs[..],
            &scan::Opts {
                include_binary: false,
                include_all: true,
            },
        );
        assert_debug_snapshot!(normalize_matches(&res.unwrap()));
    }
}
