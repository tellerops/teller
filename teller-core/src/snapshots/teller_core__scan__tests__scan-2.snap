---
source: teller-core/src/scan.rs
expression: res
---
Ok(
    [
        Match {
            path: "fixtures/config.yml",
            position: Some(
                (
                    3,
                    11,
                ),
            ),
            offset: 32,
            query: KV {
                value: "hashicorp",
                key: "key1",
                from_key: "key1",
                path: Some(
                    PathInfo {
                        id: "/some/path",
                        path: "/some/path",
                    },
                ),
                provider: Some(
                    ProviderInfo {
                        kind: Inmem,
                        name: "test",
                    },
                ),
                meta: None,
            },
        },
    ],
)
